flink-optimizer/src/main/java/org/apache/flink/optimizer/plantranslate/JobGraphGenerator.java

```diff
@@ -480,6 +482,32 @@ public void postVisit(PlanNode node) {
 			if (node instanceof SourcePlanNode || node instanceof NAryUnionPlanNode || node instanceof SolutionSetPlanNode) {
 				return;
 			}
+
+			// if this is a blocking shuffle vertex, we add one IntermediateDataSetID to its predecessor and return
+			if (node instanceof SinkPlanNode) {
+				Object userCodeObject = node.getProgramOperator().getUserCodeWrapper().getUserCodeObject();
+				if (userCodeObject instanceof BlockingShuffleOutputFormat) {
+					PlanNode precedentNode = node.getInputs().iterator().next().getSource();
+					JobVertex precedentVertex;
+					if (this.vertices.containsKey(precedentNode)) {
```

Comment:
I think we should not use `this` in the detail process in the new format except in the constructor.

Reply:
I found ``this`` shows up all over the ``JobGraphGenerator``, maybe a refactor needed later.

Reply:
Yes, we have some history inconsistent issues. But it should not bring that in new codes. For the history codes we might solve in a separate hotfix or just keep that.